src/txHashBuilder.c

```diff
@@ -368,23 +448,77 @@ void txHashBuilder_addOutput_token(
 	addToken(builder, assetNameBuffer, assetNameSize, amount,
 	         TX_HASH_BUILDER_IN_OUTPUTS_TOKEN,
 	         TX_HASH_BUILDER_IN_OUTPUTS_ASSET_GROUP,
-	         includeDatumHash ? TX_HASH_BUILDER_IN_OUTPUTS_DATUM_HASH : TX_HASH_BUILDER_IN_OUTPUTS,
+	         TX_HASH_BUILDER_IN_OUTPUTS_ASSET_GROUP,
 	         CBOR_TYPE_UNSIGNED);
 }
 
+static void txHashBuilder_assertCanLeaveOutputTopLevelData(tx_hash_builder_t *builder)
+{
+	_TRACE("state = %d, remainingOutputs = %u", builder->state, builder->multiassetData.remainingAssetGroups);
+
+	switch (builder->state) {
+	case TX_HASH_BUILDER_IN_OUTPUTS_TOP_LEVEL_DATA:
+	case TX_HASH_BUILDER_IN_OUTPUTS_ASSET_GROUP:
+		ASSERT(builder->multiassetData.remainingAssetGroups == 0);
+		break;
+
+	default:
+		ASSERT(false);
+	}
+	ASSERT(builder->remainingOutputs == 0);
+}
+
 void txHashBuilder_addOutput_datumHash(
         tx_hash_builder_t* builder,
         const uint8_t* datumHashBuffer, size_t datumHashSize
 )
 {
 	ASSERT(datumHashSize == OUTPUT_DATUM_HASH_LENGTH);
-	ASSERT(builder->state == TX_HASH_BUILDER_IN_OUTPUTS_DATUM_HASH);
+	txHashBuilder_assertCanLeaveOutputTopLevelData(builder);
 
 	{
 		BUILDER_APPEND_CBOR(CBOR_TYPE_BYTES, datumHashSize);
 		BUILDER_APPEND_DATA(datumHashBuffer, datumHashSize);
 	}
-	builder->state = TX_HASH_BUILDER_IN_OUTPUTS;
+	builder->state = TX_HASH_BUILDER_IN_OUTPUTS_DATUM_HASH;
+}
+
+void txHashBuilder_addOutput_datumOption(tx_hash_builder_t *builder, datum_option_type_t datumOption, const uint8_t *buffer,
+        size_t bufferSize)
+{
+	ASSERT(datumOption ? bufferSize < BUFFER_SIZE_PARANOIA : bufferSize == OUTPUT_DATUM_HASH_LENGTH); //??
+	txHashBuilder_assertCanLeaveOutputTopLevelData(builder);
+
+	//   datum_option = [ 0, $hash32 // 1, data ]
+	//   Array(2)[
+	//     Unsigned[0]
+	//     Bytes[buffer]
+	//   ]
+
+	BUILDER_APPEND_CBOR(CBOR_TYPE_ARRAY, 2);
+	{
+		BUILDER_APPEND_CBOR(CBOR_TYPE_UNSIGNED, datumOption);
+	}
+	{
+		BUILDER_APPEND_CBOR(CBOR_TYPE_BYTES, bufferSize);
+		BUILDER_APPEND_DATA(buffer, bufferSize);
+	}
+	builder->state = TX_HASH_BUILDER_IN_OUTPUTS_DATUM_OPTION;
+}
+
+
+void txHashBuilder_addOutput_referenceScript(tx_hash_builder_t *builder, const uint8_t *scriptBuffer, size_t bufferSize)
```

Comment:
We need functions that will allow to add chunks of the script repeatedly. So the first function will only add the tag (6.24) and total bytestring length (the tx hash builder state machine should make sure it can only be called once) and then another function will add bytes, and can be called repeatedly.

Reply:
An analogous pair of fucntions for inline datum.